DevCycle.SDK.Server.Local/Api/EventQueue.cs

```diff
@@ -62,127 +68,92 @@ public void RemoveFlushedEventsSubscriber(EventHandler<DVCEventArgs> flushedEven
-                        {
-                            // Add non-retryable payloads to list of "completed" so that they are removed from queue
-                            completedRequests.Add(batch);
-                        }
-
-                        throw error;
-                    }
-
-                    batchQueue.Clear();
-
-                    logger.LogDebug("DVC Flushed {EventCount} Events, for {UserEventBatch} Users", eventCount,
-                        userEventBatch.Count);
-                    eventArgs.Success = true;
-                    OnFlushedEvents(eventArgs);
-                }
-                catch (DVCException e)
-                {
-                    if (e.IsRetryable())
+                    if (res.StatusCode != HttpStatusCode.Created)
                     {
-                        logger.LogError(
-                            "Error publishing events, retrying, status {status}, body: {response}", e.HttpStatusCode,
-                            response?.Content);
-                        ScheduleFlushWithDelay(true);
+                        logger.LogError($"Error publishing events, status: ${res.StatusCode}, body: ${res.Content}");
+                        localBucketing.OnPayloadFailure(this.environmentKey, flushPayload.PayloadID, (int)res.StatusCode >= 500);
+                        flushResultEvent.Success = false;
+                        flushResultEvent.Error = new DVCException(res.StatusCode, new ErrorResponse(res.ErrorMessage ?? ""));
                     }
                     else
                     {
-                        logger.LogError(
-                            "DVC Events were invalid and have been dropped, status {status}, body: {response}",
-                            e.HttpStatusCode,
-                            response?.Content);
+                        logger.LogDebug($"DVC Flushed ${eventCount} Events, for ${flushPayload.Records.Count} Users");
+                        localBucketing.OnPayloadSuccess(this.environmentKey, flushPayload.PayloadID);
                     }
-
-                    eventArgs.Success = false;
-                    eventArgs.Error = e;
-                    OnFlushedEvents(eventArgs);
                 }
-                catch (Exception e)
+                catch (DVCException ex)
                 {
-                    logger.LogError(
-                        "Something went wrong flushing events, retrying, {message}", e.Message);
-                    ScheduleFlushWithDelay(true);
-                    eventArgs.Success = false;
-                    eventArgs.Error =
-                        new DVCException(HttpStatusCode.InternalServerError, new ErrorResponse(e.Message));
-                    OnFlushedEvents(eventArgs);
+                    logger.LogError($"DVC Error Flushing Events response message: ${ex.Message}");
+                    localBucketing.OnPayloadFailure(this.environmentKey, flushPayload.PayloadID, true);
```

Comment:
Mutex/Semaphore here too

Reply:
I'll create a ticket for handling multi-threading issues. We should write some tests that fail if we're not handling this correctly, and then add it in.